### PR TITLE
Add policy factory overloads for pooled instance registration

### PR DESCRIPTION
**Pull request by zms9110750**
Hi, this was created with AI assistance to address issue #6. However, I'm not very familiar with this project, so I'd appreciate a review to confirm everything is correct.

---
Add new `PooledInstancePerLifetimeScope(Func<IComponentContext, IPooledRegistrationPolicy<TLimit>>)` and `PooledInstancePerMatchingLifetimeScope(Func<IComponentContext, IPooledRegistrationPolicy<TLimit>>, params object[])` overloads that accept a factory delegate, allowing the pooling policy to be resolved from the container at registration time rather than supplied as an already-built instance.

This enables policies with container-managed dependencies (e.g., a cache-backed policy that resolves `IMemoryCache` from DI).

### Design

The factory is called **once** when the pool is built, and the resulting policy instance is shared between the pool's create/return path and the get-side resolve path via a new internal `PooledInstanceContext<T>` type. This preserves the behavior of the existing `IPooledRegistrationPolicy<T>` overloads — a single policy object handles both `Get` and `Return`.

The factory is passed as a typed constructor argument to the activator (not via registration metadata), consistent with the existing overloads.

### Changes

- **`PoolActivator<T>`**: new constructor accepting `(Service, Func<IComponentContext, IPooledRegistrationPolicy<TLimit>>)`; pipeline calls factory once and wraps the pool + policy in `PooledInstanceContext<T>`
- **`PoolGetActivator<T>`**: new constructor accepting only `PoolService`; extracts shared policy from `PooledInstanceContext<T>` at resolve time (no per-resolve factory invocation)
- **`PooledInstanceContext<T>`**: new internal type that bundles the resolved pool and policy for sharing between activators
- **`RegistrationExtensions`**: 2 new public overloads — `PooledInstancePerLifetimeScope(policyFactory)` and `PooledInstancePerMatchingLifetimeScope(policyFactory, tags)`; 1 private `RegisterPooled(policyFactory, tags)` method

### Tests

9 new tests covering:
- Happy path registration and resolution
- Policy resolved from container (dependency injection)
- Cross-scope instance reuse
- Factory receives `IComponentContext` with access to registered services
- Matching lifetime scope tag support
- Non-singleton policy sharing (factory called exactly once)
- Null argument guards for both overloads
- Concurrent usage

### CODE REVIEW

**Comment by tillig** [comment]
Thank you for putting this together — and for sticking with the conversation while we worked out exactly what you were after. You were right that there's a real gap here, and I appreciate the effort you put into the implementation and tests.

I'd like to help get this merged, but I think we'll get there faster if we narrow the scope a bit. Here's how I'm seeing it:

## The part I'd love to take

The core of your request — being able to supply the pooling policy **lazily from the container** instead of as an already-built instance — is a great fit. It's the same "give me the recipe, not the instance" pattern we use throughout Autofac, and it solves your cache-backed scenario cleanly. Concretely, that's a deferred overload of the *existing* methods:

```csharp
PooledInstancePerLifetimeScope(
    this IRegistrationBuilder<...> registration,
    Func<IComponentContext, IPooledRegistrationPolicy<TLimit>> policyFactory)
```

(plus the matching `PooledInstancePerMatchingLifetimeScope` variant). Keeping the same method names and the `IPooledRegistrationPolicy` vocabulary lets this slot right in next to the overloads that are already there.

A few implementation notes when you get to it:

- Let's invoke the factory once when the pool is built and capture the result, rather than per-resolve — keeps the allocation/perf profile in line with the existing path.
- Pass the factory through as a constructor argument to the activator (the way the current policy is) rather than via registration metadata, so it doesn't leak into the public component metadata.
- Worth a test that proves the policy can resolve another registered dependency (your cache) — that's the heart of the scenario and the thing I most want to see locked down.

## The part I'd like to set aside for now

The fully custom `ObjectPool<T>` path (`ConfigurePool`) is trickier, and I don't think we should rush it. When the pool itself is user-supplied, Autofac is no longer the thing constructing the pooled instances, which means activation and the `IPooledComponent` get/return hooks quietly stop applying. That's a real behavioral shift, and I'd want to think through whether it's something we want to own long-term before we commit to supporting it forever. If you have a concrete need there, let's discuss it in the original issue so we can design it deliberately rather than bolt it on.

## Example: your cache-backed scenario, no custom pool needed

The key insight is that the **policy** is the right place for cache-backed behavior, and as a constructor dependency, just like any other component.

Your policy decides what `Get` and `Return` mean, and it can lean on the cache for idle/pressure-based eviction:

```csharp
  public sealed class CacheBackedPolicy<T> : DefaultPooledRegistrationPolicy<T>
      where T : class
  {
      private readonly IMemoryCache _cache;

      public CacheBackedPolicy(IMemoryCache cache) => _cache = cache;

      public override T Get(IComponentContext context, IEnumerable<Parameter> parameters, Func<T> getFromPool)
      {
          // Hand back a cached instance if one is available; otherwise pull from the pool.
          if (_cache.TryGetValue(typeof(T), out T cached) && cached is not null)
          {
              _cache.Remove(typeof(T));
              return cached;
          }

          return getFromPool();
      }

      public override bool Return(T pooledObject)
      {
          // Park the instance in the cache with whatever eviction policy you like
          // (sliding expiration for idle time, size limits for pressure, etc.).
          _cache.Set(typeof(T), pooledObject, new MemoryCacheEntryOptions
          {
              SlidingExpiration = TimeSpan.FromMinutes(5),
              Size = 1,
          });

          // Returning false means "I've taken ownership of this instance" — the pool
          // won't retain it, because the cache is now managing its lifetime.
          return false;
      }
  }
```

And registration ties it together — note the cache is registered normally, and the policy factory resolves it at build time, which is exactly the ordering problem you ran into:

```csharp
  builder.RegisterType<MemoryCache>()
         .As<IMemoryCache>()
         .SingleInstance();

  builder.RegisterType<Hello>()
         .PooledInstancePerLifetimeScope(ctx =>
             new CacheBackedPolicy<Hello>(ctx.Resolve<IMemoryCache>()));
```

That gives you a generic, cache-driven pooling strategy that's resolved from the container, without Autofac having to hand off the whole `ObjectPool<T>`. The policy keeps Autofac in charge of constructing instances (so constructor injection and the `IPooledComponent` hooks keep working), while you stay in charge of retention and eviction.

## Suggested path forward

Could we trim this PR down to just the policyFactory overloads (with the existing method names) and a couple of focused tests? That's something I can review and merge quickly. The custom-pool idea can live as its own discussion, and if we land on a good design it becomes a clean follow-up PR.

None of this is a reflection on the work — it's solid, and the underlying idea is one we want. I'd just rather ship the high-value piece now and give the bigger idea the consideration it deserves. Thanks again for contributing!

**Comment by zms9110750** [comment]
Actually, I still want the `Func<IComponentContext, ObjectPool<T>>` overload.

In my view, the semantics of an object pool should be the same as a cache. I'm not satisfied with Microsoft's default `ObjectPool<T>` implementation itself.

If you only need the `Func<IComponentContext, IPooledRegistrationPolicy<T>>` part, the very first version of this commit was exactly that.

The problem was that my English isn't good, so I relied on AI to analyze the project. I didn't realize that `IPooledRegistrationPolicy` is a policy abstraction, not the pool itself. That's why in the first commit, I didn't wrap the policy with a default pool.

If you don't accept the `Func<IComponentContext, ObjectPool<T>>` overload, I'll be disappointed. But I can still help you get the `Func<IComponentContext, IPooledRegistrationPolicy<T>>` part merged. As a DI container, the beauty of it is that you can register open generics for anything — including this policy.

Still, the same issue remains: the content was created by AI and needs review.

**Comment by tillig** [comment]
That's really helpful context — thank you for being so candid about how the PR came together. And to be clear up front: **I'm not saying no to `Func<IComponentContext, ObjectPool<T>>`.** I'm saying it needs more design than this PR worked through, and I'd rather get there deliberately than merge something we'll all have to live with forever.

Your motivation is completely legitimate. "I don't want Microsoft's default `ObjectPool<T>`, I want pool semantics that behave like a cache" is a real, reasonable goal, and "register it as an open generic" is exactly the kind of thing a DI container *should* make easy. So the idea has my interest.

The reason I want to slow down is that handing Autofac a fully custom `ObjectPool<T>` changes some load-bearing assumptions, and the current implementation doesn't yet have...

- **Instance construction.** Today Autofac builds the pooled instances (constructor injection, parameters, the registration's activator). A user-supplied pool calls its own `Get()`, so who constructs the instance — and how does it get its dependencies injected? Right now that path quietly bypasses Autofac's activation.
- **`IPooledComponent` hooks.** `OnGetFromPool` / `OnReturnToPool` stop firing on the custom-pool path. That's a silent behavioral change for anyone relying on them.
- **Ownership and disposal.** With the default provider we know who disposes instances that aren't retained. With an external pool, the ownership contract is undefined — we need to decide and document who disposes what, and when.
- **Pool lifetime.** The pool is registered at the root scope. If it's resolved from the container, we need to be explicit about its lifetime, disposal, and thread-safety guarantees.
- **Open generics.** This is the scenario you actually care about, and it's the one the PR doesn't exercise at all — there are no tests proving an open-generic custom pool resolves and behaves correctly.
- **Test coverage.** The `ConfigurePool` path currently has no tests, so none of the above is pinned down.

None of these are dealbreakers — they're just open questions, and I don't think it's fair to you or to the project to merge the custom-pool API before we've answered them. That's design work worth doing in its own issue, where we can talk through the contract before writing more code.

So here's what I'd suggest, and I think it's genuinely good news for you:

1. **Let's land the `Func<IComponentContext, IPooledRegistrationPolicy<T>>` overload now.** You've offered to help get that part merged — that would be fantastic, and it's immediately useful. It's a clean, low-risk first step.
2. **Let's keep the custom-pool idea alive in [#6](https://github.com/autofac/Autofac.Pooling/issues/6)** (or a fresh issue), where we work through the bullets above together. Once we agree on the contract, the implementation becomes a focused follow-up PR — and you'll have shipped the foundation it builds on.

This isn't the door closing on the custom pool; it's us taking it one solid step at a time so each piece is something we're confident owning long-term. I'd really like to get your name on both. And no worries at all about the AI-assisted parts — that's exactly what review is for. Thanks for pushing on this.

**Comment by zms9110750** [comment]
Thank you so much for your detailed response.

The thorough context translated your statements with remarkable precision.
Your kindness and courtesy also felt genuinely warm.

I believe this commit should now be exactly what you're looking for.
I'm excited to think through the next steps with you.

File: src/Autofac.Pooling/PoolGetActivator.cs [hunk 1]
```diff
@@ -39,6 +55,7 @@ public void ConfigurePipeline(IComponentRegistryServices componentRegistryServic
         pipelineBuilder.Use(PipelinePhase.Activation, (ctxt, next) =>
         {
             var pool = (ObjectPool<TLimit>)ctxt.ResolveService(_poolService);
+            var policy = _policyFactory?.Invoke(ctxt) ?? _registrationPolicy!;
```
Comment (tillig) [inline src/Autofac.Pooling/PoolGetActivator.cs]: The factory is invoked here on **every resolve**, which means the policy on the get-side can be a different instance from the one `PoolActivator` baked into the pool's `Return` path. With the existing `IPooledRegistrationPolicy` overload, a single policy object handles both `Get` and `Return`; here, unless the factory happens to return a singleton, `Get` and `Return` run against different policy instances (and different `IComponentContext` scopes).

Any stateful policy breaks under that split — our own `PoolTrackingPolicy` test helper, which increments on `Get` and decrements on `Return`, would have its counter divided across two objects. The current tests don't surface this because they all register the policy via `RegisterInstance(...)`, which masks it.

Please invoke the factory **once** when the pool is created, capture the resulting policy, and share that single instance between the pool's `Return` and the get-side `Get`. That also removes the per-resolve container resolve from the hot path.

File: src/Autofac.Pooling/PoolActivator.cs [hunk 2]
```diff
@@ -42,15 +61,31 @@ public void ConfigurePipeline(IComponentRegistryServices componentRegistryServic
     {
         pipelineBuilder.Use(PipelinePhase.Activation, (context, next) =>
         {
-            // Get a reference to the actual lifetime scope.
-            var scope = context.Resolve<ILifetimeScope>();
+            if (_policyFactory is not null)
```
Comment (tillig) [inline src/Autofac.Pooling/PoolActivator.cs:64]: This is the other half of the issue on the get-side — the factory is invoked here to size/build the pool, *and* separately in `PoolGetActivator`. The same policy instance produced here needs to be the one used on the get-side, so the two activators don't independently call the factory. If we capture the policy once (ideally at registration time, handed to both activators), this branch and the `PoolGetActivator` branch collapse back toward the existing single-policy code path.

File: src/Autofac.Pooling/PooledInstanceTracker.cs [hunk 3]
```diff
@@ -39,6 +39,11 @@ public TPooledObject Instance
     /// </summary>
     public void Dispose()
     {
+        if (Instance is IPooledComponent poolAwareComponent)
+        {
+            poolAwareComponent.OnReturnToPool();
```
Comment (tillig) [inline src/Autofac.Pooling/PooledInstanceTracker.cs]: Please split this out of the PR. Moving `OnReturnToPool()` from `AutofacPooledObjectPolicy.Return` into `PooledInstanceTracker.Dispose` isn't needed for the factory overload, and it touches the core return path that **every** existing pooling registration goes through — so it widens the regression surface here for no reason.

It also looks like a behavioral change, not just a move: in the old location `OnReturnToPool()` fired inside the pool's `Return`; in `Tracker.Dispose` it fires unconditionally just before `_pool.Return(...)`. I want to be sure we're not changing *when* the hook fires relative to a policy returning `false`. If there's a real bug being fixed, let's do it as its own PR with a test that pins the ordering; otherwise, please revert it here (including the removed block in `AutofacPooledObjectPolicy.Return`).

File: src/Autofac.Pooling/RegistrationExtensions.cs [hunk 4]
```diff
@@ -378,4 +488,115 @@ private static void RegisterPooled<TLimit, TActivatorData, TSingleRegistrationSt
 
         callback.Callback = newCallback;
     }
+
+
```
Comment (tillig) [inline src/Autofac.Pooling/RegistrationExtensions.cs]: There are two consecutive blank lines here. With `AnalysisMode=AllEnabledByDefault` and warnings-as-errors in Release, StyleCop SA1507 will fail CI on this. Please run `dotnet msbuild ./default.proj` in Release locally to confirm the whole thing builds clean before pushing.

File: src/Autofac.Pooling/RegistrationExtensions.cs [hunk 5]
```diff
@@ -138,6 +138,58 @@ public static IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationSt
...
+    /// <typeparam name="TActivatorData">Activator data type.</typeparam>
+    /// <typeparam name="TSingleRegistrationStyle">Registration style.</typeparam>
+    /// <param name="registration">The registration.</param>
+    /// <param name="policyFactory">
+    /// A factory that returns the <see cref="IPooledRegistrationPolicy{TLimit}"/> to use for this registration.
+    /// Invoked during resolve with access to the current <see cref="IComponentContext"/>.
+    /// </param>
+    /// <returns>The registration builder.</returns>
+    public static IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle>
+           PooledInstancePerLifetimeScope<TLimit, TActivatorData, TSingleRegistrationStyle>(
```
Comment (tillig) [inline src/Autofac.Pooling/RegistrationExtensions.cs:171]: The API shape is exactly right here — same method name, same `IPooledRegistrationPolicy` vocabulary, sitting right alongside the existing overloads. 👍

File: test/Autofac.Pooling.Test/FactoryOverloadTests.cs [hunk 6]
```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Autofac Project. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Autofac.Pooling.Tests.Common;
+using Xunit;
+
+namespace Autofac.Pooling.Test;
+
+public class FactoryOverloadTests
```
Comment (tillig) [inline test/Autofac.Pooling.Test/FactoryOverloadTests.cs:12]: Good coverage of the happy path, and `PolicyFactory_PolicyFactoryReceivesComponentContext` nicely proves a dependency can be resolved at factory time. Please close these gaps:

- **A non-singleton, stateful policy test** — register the policy so the factory returns a *fresh* instance per call (e.g. `ctx => new PoolTrackingPolicy<...>()`), then assert `Get`/`Return` accounting stays consistent across a scope. This is the test that catches the per-resolve issue flagged in `PoolGetActivator`, so it's the key addition.
- **Null-argument tests** for the new guards on both overloads.
- A **concurrency** test consistent with the existing `ConcurrencyTests` suite, since the pool is shared at root scope.
- `PolicyFactory_PolicyAcceptsAllReturns` doesn't assert anything meaningful — it just resolves twice and disposes. Please give it a concrete assertion or drop it.

**Review by tillig — CHANGES_REQUESTED** [review]
This is a big step in the right direction — thank you for reworking it. Dropping the custom `ObjectPool<T>` path and scoping down to `policyFactory` overloads on the existing `PooledInstancePerLifetimeScope` / `PooledInstancePerMatchingLifetimeScope` methods is exactly the shape we were hoping for. The API naming and vocabulary are consistent with what's already there, the metadata leak is gone, the factory is passed as a typed constructor argument, and there's no dead code this time. The direction is right and this is close.

I'm requesting changes on the following — I'd like all of these addressed before merge:

1. **Resolve the policy once and share that single instance** between pool creation (`PoolActivator`) and the get-side (`PoolGetActivator`). Right now the factory is invoked per-resolve on the get-side and separately when the pool is built, so a non-singleton policy ends up with `Get` and `Return` running on different instances. A stateful policy would break, and the current tests don't catch it because they all use `RegisterInstance`.
2. **Revert (or split out) the `OnReturnToPool` move** between `AutofacPooledObjectPolicy` and `PooledInstanceTracker` — it's unrelated to this feature, touches the shared return path, and may quietly change when the hook fires.
3. **Add a non-singleton/stateful policy test** (this locks down #1), plus null-argument tests, and a concurrency case. Give `PolicyFactory_PolicyAcceptsAllReturns` a real assertion or remove it.
4. **Clean up the double blank line** and confirm a green Release build via `dotnet msbuild ./default.proj`.
5. **Update the PR title and description** — both still describe the removed `ObjectPool<T>` design. The title ("Add ObjectPool<T> factory overload...") should reflect what the PR now does, e.g. *"Add policy factory overloads for pooled instance registration"*, since it becomes the squash-merge commit message and release-note entry. The body references a `DelegatingPooledRegistrationPolicy<T>` type and an `ObjectPool<TLimit>` factory overload, neither of which exists in the current diff — please update it to describe the `Func<IComponentContext, IPooledRegistrationPolicy<TLimit>>` overloads that are actually here.

Once the policy-instance handling is sorted and the unrelated refactor is out, I'm happy to merge. Really appreciate how you've engaged with the feedback on this — the design is in good shape now. And as discussed, the custom-pool idea stays alive in #6 as a deliberate follow-up.

**Review by tillig — APPROVED** [review]
This looks great. Thanks for the contribution! We can continue the custom object pool design discussion in #6 .
